Sources/OpenCombine/Publishers/TransformingInner.swift

```diff
@@ -0,0 +1,70 @@
+//
+//  TransformingInner.swift
+//
+//  Created by Eric Patey on 26.08.2019.
+//
+
+internal class TransformingInner<Upstream: Publisher, Downstream: Subscriber>
```

Comment:
To be honest, I'm not in favor of this `TransformingInner` class, sorry.

It seems to use stored properties like `shouldProxySubscription` and `description` just to mimic Combine, while not adding any value beyond that, but possibly introducing overhead.

My suggestion would be to leave `Map` and `TryMap` as is, and implement `Scan` and `TryScan` using an ad-hoc subscription class. Yep, it's a little bit of duplication, but least at it doesn't make things more complex.

Reply:
@broadwaylamb, thanks for the review and feedback. I appreciate it.

I'd like to discuss it a little bit further to make sure I'm aligned specifically with the sorts of things that you consider unnecessary complexity.

I think it's worth the extra effort, because I'm motivated very strongly by code sharing and a D.R.Y. approach. If done properly, I think sharing as much code as possible between operators will have massive benefit and allow us to deliver a more consistently correct implementation.

That said, I agree that some of my shared implementation was a bit lazy. Specifically, some of my use of stored properties as you pointed out. I think the generalized concern is paying any memory/runtime expense to support `.tryScan` if I only ever use `.scan`.

My use of `shouldProxySubscription` was a great example of failing that test. Though it didn't incur a memory penalty, my implementation of `description` in the base class did cause unneeded runtime cost during init time.

I've made another pass as the shared class to get rid of all of those stored properties and made it simpler to consume. It now deals with all of the following differences between the throwing and non-throwing variants without leaking any cost/complexity between the two:
- Conforming to `Subscription` in the throwing variant and passing `self` 
- Differences in the implementation of `Subscriber.receive(subscription:)` to pass the proper `Subscription` downstream: the upstream subscription for non-throwing variant and `self` for the throwing variant. 
- Adapting `Subscribers.Completion<Upstream.Failure>` to `Subscribers.Completion<Error>` for the throwing variant.

The more I work on this stuff, the more I am convinced that there are many non-obvious Apple behaviors that we need to match. Because of this, I think it would be a mistake to have similar but copied and likely divergent code across the project. Here's one example that we're already soaking in. Apple's behavior if `.receive(completion:)` is called multiple times is non-obvious. They allow multiple `.failure`'s to go downstream, but they do NOT allow multiple `.finished`'s to go downstream. It's not crazy complex, but it's subtle and non-obvious. As currently implemented, this is tested properly for `.map`, but not for `.tryMap`. In fact, the code is broken for `.tryMap`. Simply add a copy of `testError` that uses `tryMap` rather than `map`. My point is that we already have inconsistencies that have caused bugs. Properly tested shared code would benefit ALL consumers of that shared code.

So, that all said, I'd really appreciate it if you took another pass a the code now that I've incorporated your feedback.